### PR TITLE
parquet: Add support for user-provided metadata loaders

### LINKED ISSUE #12592: parquet: Add support for user-provided metadata loaders
### Is your feature request related to a problem or challenge?

This allows users to implement #12547 (caching metadata, especially the Page Index) themselves without reimplementing all of `ParquetOpener` and `ParquetExec`

### Describe the solution you'd like

Add a method to the object returned by `ParquetFileReaderFactory` that loads metadata, so that users who overwrite the `DefaultParquerFileReaderFactory` with `SessionState::register_file_format(..., true)` can provide their own implementation.

### Describe alternatives you've considered

Adding caching support directly in Datafusion. I tried in #12548 but it went nowhere.

Support for caching directly in Datafusion can be added later, based on this feature.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by progval**
## Which issue does this PR close?

Closes #12592.

## Rationale for this change

This allows users to, for example, cache the Page Index so it does not need to be parsed every time we open the file.

If have a demo here: https://gitlab.softwareheritage.org/swh/devel/swh-provenance/-/merge_requests/182 , the key thing being a `CachingParquetFormatFactory`/`CachingParquetFormat` pair that acts like `ParquetFormatFactory`/`ParquetFormat` but they call `ParquetExecBuilder::with_parquet_file_reader_factory` to a file reader factory that keeps a pool of readers (keyed by file path). It gives a significant improvement on the `time_elapsed_opening` metric:

> statistics_eval_time=775.283531ms, row_pushdown_eval_time=191ns, time_elapsed_scanning_until_data=16.463873ms, time_elapsed_processing=401.16043928s, **time_elapsed_opening=473.912060225s**, page_index_eval_time=24.19564ms, **metadata_load_time=410.282903954s**, bloom_filter_eval_time=62.038167937s, time_elapsed_scanning_total=16.516948ms

to 

> row_pushdown_eval_time=191ns, time_elapsed_scanning_until_data=1.143893882s, time_elapsed_processing=34.501308377s, **time_elapsed_opening=67.214927741s**, page_index_eval_time=32.002383ms, **metadata_load_time=3.411246ms**, bloom_filter_eval_time=66.495163165s, statistics_eval_time=1.255058963s, time_elapsed_scanning_total=1.143978104s

## What changes are included in this PR?

* Renamed `ParquetFileReader` struct to `DefaultParquetFileReader`
* Add new `ParquetFileReader` trait that extends `AsyncFileReader` with a `load_metadata` method.
* Call it from `<ParquetOpener as FileOpener>::open`

## Are these changes tested?

Not within the repo. Should I add a new module `datafusion-examples/` adapted from my demo above.

## Are there any user-facing changes?

Breaking change for any user who implements `ParquetFileReaderFactory`.

### CODE REVIEW

**Comment by alamb** [comment]
I plan to review this later today

**Review by alamb — COMMENTED** [review]
Thank you for this contribution @progval 

I went through this change and I am not sure what can not be done with the existing APIs. The example you cite:

> This allows users to, for example, cache the Page Index so it does not need to be parsed every time we open the file.

I believe the existing [`AsyncReader::get_metadata`](https://docs.rs/parquet/latest/parquet/arrow/async_reader/trait.AsyncFileReader.html#tymethod.get_metadata) already permits this and it is actually already done by the advanced_parquet_index.rs example

https://github.com/apache/datafusion/blob/91c8a47f2001b29bb4fdf6f41e18a65bbf0752de/datafusion-examples/examples/advanced_parquet_index.rs#L574-L582

One thing that I found very unclear when working with this code at first, is that if the ParquetMetaData returned by  [`AsyncReader::get_metadata`](https://docs.rs/parquet/latest/parquet/arrow/async_reader/trait.AsyncFileReader.html#tymethod.get_metadata) does not *already* have the page and offset index loaded the ArrowReader will in fact try and read the index information with new requests ([doc](https://docs.rs/parquet/latest/parquet/arrow/arrow_reader/struct.ArrowReaderMetadata.html#notes-2)).  

When building the cache in the example, it was important to ensure that the page indexes were read as part of the initial load here:

https://github.com/apache/datafusion/blob/91c8a47f2001b29bb4fdf6f41e18a65bbf0752de/datafusion-examples/examples/advanced_parquet_index.rs#L408-L415

Perhaps your system needs to do something similar?

**Comment by alamb** [comment]
I looked a bit at https://gitlab.softwareheritage.org/swh/devel/swh-provenance/-/merge_requests/182/diffs

I wonder if you could cache the Arc<ParquetMetadata> rather than the `ArrowReaderMetadta` 🤔  (I find the maze of similarly named "Reader/loader/Arrow/metadata" structures hard to keep straight sometimes

**Comment by progval** [comment]
I agree, it's confusing.

I actually started from [advanced_parquet_index.rs](https://github.com/apache/datafusion/blob/91c8a47f2001b29bb4fdf6f41e18a65bbf0752de/datafusion-examples/examples/advanced_parquet_index.rs#L574-L582), but it appeared to be insufficient.

But reading this again, I agree with you that it should be. I'll give it another shot.

**Comment by progval** [comment]
I remembered. The issue is that in this code:

https://github.com/apache/arrow-rs/blob/62825b27e98e6719cb66258535c75c7490ddba44/parquet/src/arrow/async_reader/mod.rs#L212-L228

`ArrowReaderMetadata::load_metadata` does two expensive things:

1. call `AsyncFileReader::get_metadata()` to get a `ParquetMetadata`, which we can cache as per `advanced_parquet_index.rs`
2. instantiate and use [`parquet::MetadataLoader`](https://docs.rs/parquet/latest/parquet/arrow/async_reader/struct.MetadataLoader.html) to get **another** `ParquetMetadata` from the first one. And this new `ParquetMetadata` isn't cachable without my patch.

in essence, it's enriching the `metadata` struct by doing: `let metadata = MetadataLoader::new(..., metadata).load_page_index(...).finish()`

**Comment by progval** [comment]
To give an idea of the difference. Metrics with the new `load_metadata` cached:

> time_elapsed_opening=272.811692949s, time_elapsed_processing=118.675115374s, time_elapsed_scanning_total=2.530368545s, time_elapsed_scanning_until_data=2.530322168s

 Metrics with only `get_metadata cached`:

> time_elapsed_opening=165.253279952s, time_elapsed_processing=36.211909412s, time_elapsed_scanning_total=605.765021ms, time_elapsed_scanning_until_data=605.726082ms

File: datafusion/core/src/datasource/physical_plan/parquet/reader.rs [hunk 1]
```diff
@@ -57,9 +58,49 @@ pub trait ParquetFileReaderFactory: Debug + Send + Sync + 'static {
         file_meta: FileMeta,
         metadata_size_hint: Option<usize>,
         metrics: &ExecutionPlanMetricsSet,
-    ) -> datafusion_common::Result<Box<dyn AsyncFileReader + Send>>;
+    ) -> datafusion_common::Result<Box<dyn ParquetFileReader>>;
 }
 
+/// [`AsyncFileReader`] augmented with a method to customize how file metadata is loaded.
```
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/parquet/reader.rs:64]: I wonder if we can use one of the (many!) already existing APIs /traits for loading the metadata. The number of layers of abstraction are already pretty mind boggling. Adding yet another one seems unweildy. Or maybe we could just change the code not to call the loader if it has metadata 🤔

File: datafusion/core/src/datasource/physical_plan/parquet/reader.rs [hunk 2]
```diff
@@ -57,9 +58,49 @@ pub trait ParquetFileReaderFactory: Debug + Send + Sync + 'static {
...
+pub trait ParquetFileReader: AsyncFileReader + Send + 'static {
+    /// Returns a [`AsyncFileReader`] trait object
+    ///
+    /// This can usually be implemented as `Box::new(*self)`
+    fn upcast(self: Box<Self>) -> Box<dyn AsyncFileReader + 'static>;
```
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/parquet/reader.rs:69]: This method is unfortunate, but I also could not figire out  a way without it

**Review by alamb — COMMENTED** [review]
> I remembered / rediscovered the issue. It's in this code:

Right -- I agree this is confusing.

I think you can work around it by ensuring that the `ParquetMetadata` that is passed back already has the page_index loaded -- so ` MetadataLoader::new()....load_page_index(true).build` is a noop

However, given I hit the basically the exact same thing when writing the advanced parquet index, and now you have hit it as well, I think changing / making it clearer how to handle this is in order

I actually like the idea of a PR / change in behavior to stop this automatic fetching from happening (as presumably the reason for using a custom loader is to avoid having to fetch the metadata!)

Maybe we can just change the code to not fetch the metadata (and just use what is passed back)

**Comment by alamb** [comment]
I wonder if we could change the "automatically load page index if needed" to "error if page index is needed but it is not loaded" 🤔  That might be a less surprising behavior

**Comment by alamb** [comment]
Thanks @progval  -- I will find time to review this more carefully tomorrow. 

My primary concerns are with the API being even more complicated to use correctly than the current one (the upcast thing is confusing and "feels" wrong to me, as does the need for *another* API in DataFusion that wraps an already complicated API in parquet.

I am hopeful we can come up with something more elegant

**Comment by progval** [comment]
> (For some reason it's a bit slower even though it does cache as it should. I'll investigate in two weeks)

Actually, it's not slower. I must have had a noisy benchmark or bad luck eyeballing the mean request time.

**Pull request closed without merging**